File: src/main/java/com/raychenon/leetcode/tree/SubtreeOfAnotherTree.java
package com.raychenon.leetcode.tree;

/**
 * User: raychenon
 * Date: 8/3/2020
 * https://leetcode.com/problems/subtree-of-another-tree
 */
public class SubtreeOfAnotherTree {

    public boolean isSubtree(TreeNode s, TreeNode t) {
        return traverse(s, t);
    }

    public boolean equals(TreeNode x, TreeNode y) {
        if (x == null && y == null)
            return true;
        if (x == null || y == null)
            return false;
        return x.value == y.value && equals(x.left, y.left) && equals(x.right, y.right);
    }

    public boolean traverse(TreeNode s, TreeNode t) {
        return s != null && (equals(s, t) || traverse(s.left, t) || traverse(s.right, t));
    }

}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/tree/SubtreeOfAnotherTree.kt
package com.raychenon.kotlin.leetcode.tree

import com.raychenon.leetcode.tree.TreeNode

/**
 * User: raychenon
 * Date: 11/3/2020
 * https://leetcode.com/problems/subtree-of-another-tree/
 */

object SubtreeOfAnotherTree {

    fun isSubtree(s: TreeNode?, t: TreeNode?): Boolean {
        return traverse(s, t)
    }

    fun equals(x: TreeNode?, y: TreeNode?): Boolean {
        if (x == null && y == null) return true
        return if (x == null || y == null) false
        else x.value === y.value && equals(x.left, y.left) && equals(x.right, y.right)
    }

    fun traverse(s: TreeNode?, t: TreeNode?): Boolean {
        return s != null && (equals(s, t) || traverse(s.left, t) || traverse(s.right, t))
    }

}